pkg/api/customization/workload/workload.go

```diff
@@ -36,33 +38,108 @@ func (a ActionWrapper) ActionHandler(actionName string, action *types.Action, ap
+		clusterName := a.ClusterManager.ClusterName(apiContext)
+		if clusterName == "" {
+			return httperror.NewAPIError(httperror.ServerError, fmt.Sprintf("Error rolling back %s - clusterName empty", deployment.ID))
+		}
+		clusterContext, err := a.ClusterManager.UserContext(clusterName)
+		if err != nil {
+			return httperror.NewAPIError(httperror.ServerError, fmt.Sprintf("Error rolling back %s", deployment.ID))
+		}
+		err = clusterContext.UnversionedClient.Post().Prefix("apis/extensions/v1beta1/").Namespace(namespace).Resource("deployments").Name(name).SubResource("rollback").Body(deploymentRollback).Do().Error()
 		if err != nil {
-			return httperror.NewAPIError(httperror.ServerError, fmt.Sprintf("Error updating workload %s by %s", deployment.ID, actionName))
+			return httperror.NewAPIError(httperror.ServerError, fmt.Sprintf("Error updating workload %s by %s : %s", deployment.ID, actionName, err.Error()))
 		}
 
-	case "pause", "resume":
-		data, err := convert.EncodeToMap(deployment)
-		if err == nil {
-			values.PutValue(data, !deployment.DeploymentConfig.Paused, "deploymentConfig", "paused")
-			err = update(apiContext, data, deployment.ID)
+	case "pause":
+		if deployment.Paused {
+			return httperror.NewAPIError(httperror.InvalidAction, fmt.Sprintf("Deployment %s already paused", deployment.ID))
 		}
-		if err != nil {
-			return httperror.NewAPIError(httperror.ServerError, fmt.Sprintf("Error updating workload %s by %s", deployment.ID, actionName))
+		return updatePause(apiContext, true, deployment, "pause")
+
+	case "resume":
+		if !deployment.Paused {
+			return httperror.NewAPIError(httperror.InvalidAction, fmt.Sprintf("Pause deployment %s before resume", deployment.ID))
+		}
+		return updatePause(apiContext, false, deployment, "resume")
+	}
+	return nil
+}
+
+func fetchRevisionFor(apiContext *types.APIContext, rollbackInput *projectclient.DeploymentRollbackInput, namespace string, name string, currRevision string) string {
+	rollbackTo := rollbackInput.ReplicaSetID
+	if rollbackTo == "" {
+		revisionNum, _ := convert.ToNumber(currRevision)
+		return convert.ToString(revisionNum - 1)
+
+	}
+	data := getRevisions(apiContext, namespace, name, rollbackTo)
+	if len(data) > 0 {
+		return convert.ToString(values.GetValueN(data[0], "workloadAnnotations", "deployment.kubernetes.io/revision"))
+	}
+	return ""
+}
+
+func getRevisions(apiContext *types.APIContext, namespace string, name string, requestedID string) []map[string]interface{} {
+	data, replicaSets := []map[string]interface{}{}, []map[string]interface{}{}
+	options := map[string]string{"hidden": "true"}
+	conditions := []*types.QueryCondition{
+		types.NewConditionFromString("namespaceId", types.ModifierEQ, []string{namespace}...),
+	}
+
+	if err := access.List(apiContext, &projectschema.Version, projectclient.ReplicaSetType, &types.QueryOptions{Options: options, Conditions: conditions}, &replicaSets); err == nil {
+		for _, replicaSet := range replicaSets {
```

Comment:
can we perhaps add owner reference to the conditions? So the filter happens on the API side. Otherwise we get too much data back. If k8s can't search by annotations, can we use replicaset name somehow? assuming that it has similar name

Reply:
will check for if I can filter for ownerReference fields..  
Can use replicaset name - Currently using this for both List and Filter, will split and use replicaset ID